docs: add page for hints (#13)
ConsenSys/doc.gnark: docs/HowTo/write/hints.md

---
description: Using hints for off-circuit computations
---

# Compiler hints

> Instead of computing some value within a circuit, it is sometimes more optimal
> to compute the value off-circuit and only verify the correctness of the
> computation in a circuit. `gnark` provides a way to provide such verifiable
> "advice wires" in the way of *hints*. From the prover point of view, these are
> essentially input variables which are provided by hint function instead of the
> user.

Reviewer: ```suggestion
Instead of computing some value within a circuit, it's sometimes more optimal
to compute the value off-circuit and only verify the correctness of the
computation in a circuit. `gnark` allows you to do this through *hints*. From the prover's point of view, hints are
essentially input variables provided by a hint function instead of the
user.
```

> For example, lets consider a decomposition of an integer `a` into bits. A naive
> way to do it would be to look at the binary representation of the integer and
> extract bits from this representation, but `gnark` does not currently provide
> native bit operations. Instead, the hint function `hint.IthBit` can provide the
> bits as variables and the user must constrain these variables as a weighted sum
> which equals to `a`. In a circuit it would look like:

Reviewer: ```suggestion
For example, consider a decomposition of an integer `a` into bits. A naive
way to decompose it is to look at the binary representation of the integer and
extract bits from this representation, but `gnark` doesn't currently provide
native bit operations. Instead, the hint function `hint.IthBit` can provide the
bits as variables and the user must constrain these variables as a weighted sum
which equals to `a`. In a circuit it would look like:
```

```go
  var b []frontend.Variable
  var Σbi frontend.Variable
  base := 1
  for i := 0; i < nBits; i++ {
    b[i] = cs.NewHint(hint.IthBit, a, i)
    cs.AssertIsBoolean(b[i])
    Σbi = api.Add(Σbi, api.Mul(b[i], base))
    base = base << 1
  }
  cs.AssertIsEqual(Σbi, a)
```

> This particular method is also implemented in the frontends as `ToBinary(...)`
> interface method.
> 
> `gnark` provides a list of built-in hints. See [package
> documentation](https://pkg.go.dev/github.com/consensys/gnark/backend/hint#Function)
> for a list of built-int hint functions.

Reviewer: ```suggestion
This method is also implemented in the front end as `ToBinary(...)`
interface method.

`gnark` provides a [list of built-in hint functions](https://pkg.go.dev/github.com/consensys/gnark/backend/hint#Function).
```

> ## Implementing hint functions
> 
> In addition to built-in hint functions, the developer can also define their own
> hint functions. Any instance satisfying the
> [`hint.Function`](https://pkg.go.dev/github.com/consensys/gnark/backend/hint#Function)
> can be provided to `api.NewHint(...)` method to compute the hint value.
> Additionally, the hint function has to be provided as a [`backend.WithHints`
> backend option](https://pkg.go.dev/github.com/consensys/gnark/backend#WithHints)
> to the backend so that the backend could access the hint function.

Reviewer: ```suggestion
## Implement hint functions

You can define your own hint functions in addition to built-in hint functions.
You can provide any instance satisfying the
[`hint.Function`](https://pkg.go.dev/github.com/consensys/gnark/backend/hint#Function)
to `api.NewHint(...)` method to compute the hint value.
Additionally, you must provide the hint function as a [`backend.WithHints`](https://pkg.go.dev/github.com/consensys/gnark/backend#WithHints) option
to the back end, so the back end can access the hint function.
```

See https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/b/back-end

> `gnark` also provides a constructor
> [`hint.NewStaticHint`](https://pkg.go.dev/github.com/consensys/gnark/backend/hint#NewStaticHint)
> for constructing simple hint functions which takes constant number of inputs and
> returns constant number of outputs.

Reviewer: ```suggestion
`gnark` also provides a constructor
[`hint.NewStaticHint`](https://pkg.go.dev/github.com/consensys/gnark/backend/hint#NewStaticHint)
for constructing simple hint functions, which takes a constant number of inputs and
returns a constant number of outputs.
```
